contracts/vanchor/merkle_tree.rs

```diff
@@ -0,0 +1,106 @@
+use super::*;
+#[cfg(feature = "std")]
+use ink_storage::traits::StorageLayout;
+use ink_storage::traits::{SpreadAllocate, SpreadLayout};
+use ink_storage::Mapping;
+use poseidon::PoseidonRef;
+
+use crate::vanchor;
+use ink_prelude::vec;
+use scale::{Decode, Encode, Error, Input};
+
+pub const ROOT_HISTORY_SIZE: u32 = 100;
+
+#[derive(Default, Debug, SpreadLayout, SpreadAllocate)]
+#[cfg_attr(feature = "std", derive(StorageLayout, scale_info::TypeInfo))]
+pub struct MerkleTree {
+    pub levels: u32,
+    pub current_root_index: u32,
+    pub next_index: u32,
+    pub filled_subtrees: Mapping<u32, [u8; 32]>,
+    pub roots: Mapping<u32, [u8; 32]>,
+}
+
+impl Encode for MerkleTree {}
+
+impl Decode for MerkleTree {
```

Comment:
This doesn't seem like a right implementation of Decode, why not derive it? What is the error?